brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+```
+sdf.write(Arrow.RecordBatch values)          # write
+```
+
+Parameters:
+
+- values - an Arrow.RecordBatch containing all columns, including the index column. The schema for the values must match the schema for the SOMADataFrame.
+
+## SOMAMatrix
+
+> 🛑 **To be further specified** -- will look essentially the same as SOMADataFrame, extended with a second index dimension.
+
+In the following, `smx` is used to indicate an "instance" or "handle" to a SOMAMatrix.
+
+Summary:
+
+```
+SOMAMatrix.create(string uri, Arrow.Schema schema, string index_column_name) -> void
+
+SOMAMatrix.delete(string uri) -> void
+SOMAMatrix.exists(string uri) -> bool  # exists and is a SOMADataFrame
+SOMAMatrix.open(string uri, mode=read|write) -> smx
+smx.close()
+
+smx.shape -> (int, int)                     # length of each index dimension, always a list of length 2
+smx.ndims -> 2                              # returns a constant
+smx.schema -> Arrow.Schema                  # data schema, which must be the schema for an Arrow.RecordBatch
+smx.index_column_names -> (string, string)  # list of index column names, always a list of length 2
+
+smx.read(ids=[...]|all, column_names=[...]|all, filter=_TBD_) -> Arrow.RecordBatch
+smx.write(Arrow.RecordBatch values)          # write
+```
+
+## SOMA
+
+SOMA supports all SOMACollection functions, including access to the predefined fields (eg, `X`) using SOMACollection API. The primary difference in behavior between a SOMACollection and a SOMA is that the SOMA _should_ enforce naming and indexing conventions for the predefine fields (eg, `obs`), as define in the [SOMA data model](#soma).
+
+Summary:
+
+```
+SOMA.create(uri) -> soma
+SOMA.open(uri, mode=read|write) -> soma
+SOMA.delete(uri) -> void
+SOMA.exists(uri) -> bool   # exists and is a SOMA
+soma.close()
+soma.uns -> metadata
+
+type ValueType = SOMACollection | SOMADataFrame | SOMAMatrix
+
+soma.get(string key) -> ValueType
+soma.has(string key) -> bool
+soma.set(string key, ValueType value)
+soma.del(string key) -> void
+soma.__iterator__ -> key/value iterator for collection
+soma.__length__ -> int
+```
+
+## General Utilities
+
+Summary:
+
+```
+get_version() -> string                   # return semver-compatible version of the supported SOMA API
+get_implementation() -> string            # return the implementation name, eg, "R-tiledb"
+get_implementation_version() -> string    # return the package implementation version as a semver
+get_storage_engine() -> string            # return underlying storage engine name, eg, "tiledb"
+```
+
+### Method: get_SOMA_version
+
+This is a pre-release specification in active development. As defined by [semver](https://semver.org/), this API is defined as version `0.0.0-dev`.
+
+## Value Filters
+
+> 🛑 **To be specified** -- a string-based language which can express simple filters and filter combinations based on column values, eg, "col_name == 'value'".
```

Comment:
in R it's idiomatic to not use strings -- not
```
dplyr::filter(mtcars, cyl == 4)
```
but
```
dplyr::filter(mtcars, 'cyl == 4')
```

so it's high-level important that we not narrow down to "strings" here

let's keep it generic & say "query expressions" & that will be agnostic to the presence/absence of string quotes around whatever the query expression is

Reply:
Agreed. We just had a [similar conversation](https://github.com/TileDB-Inc/tiledbsc/pull/67#discussion_r906708148) about this very topic in the context of attempting to keep the r/py slicing APIs consistent but still idiomatic to the underlying language.

Reply:
In the next revision, I will attempt to thread the needle by specifying the operators and capabilities of the push-down filters, but will delegate the syntax to per-language specs.